Confusing error message when vocs goes wrong

In MOBO generator, if given a vocs like this:

```python
{
    "variables": {"x1": [0, 1.0], "x2": [0, 10.0]},
    "objectives": {"y1": "MINIMIZE"},
    "constraints": {"c1": ["GREATER_THAN", 0.5]},
    "constants": {"constant1": 1.0},
}
```

It would raise an error similar to the one below:

```
cls = <class 'xopt.generators.bayesian.mobo.MOBOGenerator'>, v = {'y1': 10.0}
info = ValidationInfo(config={'title': 'MOBOGenerator', 'extra_fields_behavior': 'forbid'}, context=None, data={'supports_bat..._time': None, 'custom_objective': None, 'n_interpolate_points': None, 'n_candidates': 1}, field_name='reference_point')

    @field_validator("reference_point")
    def validate_reference_point(cls, v, info: ValidationInfo):
>       objective_names = info.data["vocs"].objective_names
E       KeyError: 'vocs'

/usr/local/anaconda3/envs/badger-test-py310/lib/python3.10/site-packages/xopt/generators/bayesian/bayesian_generator.py:855: KeyError
```

However the correct error message should be something like "MOBO doesn't support single objective optimization problem", since we only give it a single `y1` as objective in the vocs.

--- a/xopt/generator.py
+++ b/xopt/generator.py
@@ -6,6 +6,7 @@ import pandas as pd
 from pydantic import ConfigDict, Field, field_validator
 from pydantic_core.core_schema import ValidationInfo
 
+from xopt.errors import VOCSError
 from xopt.pydantic import XoptBaseModel
 from xopt.vocs import VOCS
 
@@ -76,14 +77,14 @@ class Generator(XoptBaseModel, ABC):
     @field_validator("vocs", mode="after")
     def validate_vocs(cls, v, info: ValidationInfo):
         if v.n_constraints > 0 and not info.data["supports_constraints"]:
-            raise ValueError("this generator does not support constraints")
+            raise VOCSError("this generator does not support constraints")
         if v.n_objectives == 1:
             if not info.data["supports_single_objective"]:
-                raise ValueError(
+                raise VOCSError(
                     "this generator does not support single objective optimization"
                 )
         elif v.n_objectives > 1 and not info.data["supports_multi_objective"]:
-            raise ValueError(
+            raise VOCSError(
                 "this generator does not support multi-objective optimization"
             )
 

--- a/xopt/generators/bayesian/bayesian_generator.py
+++ b/xopt/generators/bayesian/bayesian_generator.py
@@ -858,8 +858,12 @@ class MultiObjectiveBayesianGenerator(BayesianGenerator, ABC):
 
     @model_validator(mode="after")
     def validate_reference_point(self):
+        # Note: this is called for EVERY field change and is bad for performance
+        # but we need to check in model validator to ensure vocs field has been set by field validators
         objective_names = self.vocs.objective_names
-        assert set(self.reference_point.keys()) == set(objective_names)
+        if not set(self.reference_point.keys()) == set(objective_names):
+            raise XoptError("reference point must contain all objective names in vocs")
+        return self
 
     @property
     def torch_reference_point(self):

--- a/xopt/generators/bayesian/bayesian_generator.py
+++ b/xopt/generators/bayesian/bayesian_generator.py
@@ -18,7 +18,13 @@ from botorch.sampling import get_sampler
 from botorch.utils.multi_objective import is_non_dominated
 from botorch.utils.multi_objective.box_decompositions import DominatedPartitioning
 from gpytorch import Module
-from pydantic import Field, field_validator, PositiveInt, SerializeAsAny
+from pydantic import (
+    Field,
+    field_validator,
+    PositiveInt,
+    SerializeAsAny,
+    model_validator,
+)
 from pydantic_core.core_schema import ValidationInfo
 from torch import Tensor
 
@@ -850,12 +856,10 @@ class MultiObjectiveBayesianGenerator(BayesianGenerator, ABC):
 
     supports_multi_objective: bool = True
 
-    @field_validator("reference_point")
-    def validate_reference_point(cls, v, info: ValidationInfo):
-        objective_names = info.data["vocs"].objective_names
-        assert set(v.keys()) == set(objective_names)
-
-        return v
+    @model_validator(mode="after")
+    def validate_reference_point(self):
+        objective_names = self.vocs.objective_names
+        assert set(self.reference_point.keys()) == set(objective_names)
 
     @property
     def torch_reference_point(self):

--- a/xopt/tests/generators/bayesian/test_bax.py
+++ b/xopt/tests/generators/bayesian/test_bax.py
@@ -9,7 +9,6 @@ from botorch.models import SingleTaskGP
 from botorch.models.model import ModelList
 from botorch.models.transforms import Normalize, Standardize
 
-from pydantic import ValidationError
 
 from xopt.base import Xopt
 from xopt.errors import VOCSError

--- a/xopt/tests/generators/bayesian/test_bax.py
+++ b/xopt/tests/generators/bayesian/test_bax.py
@@ -12,6 +12,7 @@ from botorch.models.transforms import Normalize, Standardize
 from pydantic import ValidationError
 
 from xopt.base import Xopt
+from xopt.errors import VOCSError
 from xopt.evaluator import Evaluator
 from xopt.generators.bayesian.bax.algorithms import (
     Algorithm,
@@ -214,5 +215,5 @@ class TestBaxGenerator:
         test_vocs = deepcopy(TEST_VOCS_BASE)
         alg = GridOptimize()
 
-        with pytest.raises(ValidationError):
+        with pytest.raises(VOCSError):
             BaxGenerator(vocs=test_vocs, algorithm=alg)

--- a/xopt/tests/generators/bayesian/test_bayesian_exploration.py
+++ b/xopt/tests/generators/bayesian/test_bayesian_exploration.py
@@ -5,6 +5,7 @@ import pytest
 from pydantic import ValidationError
 
 from xopt.base import Xopt
+from xopt.errors import VOCSError
 from xopt.evaluator import Evaluator
 from xopt.generators.bayesian.bayesian_exploration import BayesianExplorationGenerator
 from xopt.resources.testing import TEST_VOCS_BASE, TEST_VOCS_DATA, xtest_callable
@@ -104,5 +105,5 @@ class TestBayesianExplorationGenerator:
     def test_vocs_validation(self):
         test_vocs = deepcopy(TEST_VOCS_BASE)
 
-        with pytest.raises(ValidationError):
+        with pytest.raises(VOCSError):
             BayesianExplorationGenerator(vocs=test_vocs)

--- a/xopt/tests/generators/bayesian/test_bayesian_exploration.py
+++ b/xopt/tests/generators/bayesian/test_bayesian_exploration.py
@@ -2,7 +2,6 @@ from copy import deepcopy
 
 import pytest
 
-from pydantic import ValidationError
 
 from xopt.base import Xopt
 from xopt.errors import VOCSError

--- a/xopt/tests/generators/bayesian/test_bayesian_generator.py
+++ b/xopt/tests/generators/bayesian/test_bayesian_generator.py
@@ -12,6 +12,7 @@ from gpytorch.kernels import PeriodicKernel
 
 from xopt import VOCS
 from xopt.base import Xopt
+from xopt.errors import VOCSError
 from xopt.evaluator import Evaluator
 from xopt.generators.bayesian.bayesian_generator import (
     BayesianGenerator,
@@ -27,7 +28,7 @@ class PatchBayesianGenerator(BayesianGenerator):
     """
 
     supports_batch_generation: bool = True
-    #supports_multi_objective: bool = True
+    # supports_multi_objective: bool = True
     supports_single_objective: bool = True
     supports_constraints: bool = True
 
@@ -283,7 +284,7 @@ class TestBayesianGenerator(TestCase):
         )
         vocs2 = vocs.model_copy()
         vocs2.objectives = {"y1": "MINIMIZE", "y2": "MAXIMIZE"}
-        with pytest.raises(ValueError):
+        with pytest.raises(VOCSError):
             gen = MultiObjectivePatchBayesianGenerator(
                 vocs=vocs, reference_point={"y1": 0.5, "y2": 0.5}
             )
@@ -292,5 +293,5 @@ class TestBayesianGenerator(TestCase):
             vocs=vocs2, reference_point={"y1": 0.5, "y2": 0.5}
         )
         assert gen.supports_single_objective == False
-        with pytest.raises(ValueError):
+        with pytest.raises(VOCSError):
             gen.vocs = vocs

--- a/xopt/tests/generators/bayesian/test_bayesian_generator.py
+++ b/xopt/tests/generators/bayesian/test_bayesian_generator.py
@@ -292,6 +292,6 @@ class TestBayesianGenerator(TestCase):
         gen = MultiObjectivePatchBayesianGenerator(
             vocs=vocs2, reference_point={"y1": 0.5, "y2": 0.5}
         )
-        assert gen.supports_single_objective == False
+        assert not gen.supports_single_objective
         with pytest.raises(VOCSError):
             gen.vocs = vocs

--- a/xopt/tests/generators/bayesian/test_bayesian_generator.py
+++ b/xopt/tests/generators/bayesian/test_bayesian_generator.py
@@ -10,9 +10,13 @@ from botorch.models.gpytorch import GPyTorchModel
 from botorch.models.transforms import Normalize, Standardize
 from gpytorch.kernels import PeriodicKernel
 
+from xopt import VOCS
 from xopt.base import Xopt
 from xopt.evaluator import Evaluator
-from xopt.generators.bayesian.bayesian_generator import BayesianGenerator
+from xopt.generators.bayesian.bayesian_generator import (
+    BayesianGenerator,
+    MultiObjectiveBayesianGenerator,
+)
 from xopt.resources.test_functions.sinusoid_1d import evaluate_sinusoid, sinusoid_vocs
 from xopt.resources.testing import TEST_VOCS_BASE, TEST_VOCS_DATA
 
@@ -23,8 +27,12 @@ class PatchBayesianGenerator(BayesianGenerator):
     """
 
     supports_batch_generation: bool = True
-    supports_multi_objective: bool = True
+    #supports_multi_objective: bool = True
     supports_single_objective: bool = True
+    supports_constraints: bool = True
+
+
+class MultiObjectivePatchBayesianGenerator(MultiObjectiveBayesianGenerator):
     supports_constraints: bool = True
 
 
@@ -262,3 +270,27 @@ class TestBayesianGenerator(TestCase):
 
         with pytest.raises(KeyError):
             gen.train_model()
+
+    @patch.multiple(MultiObjectivePatchBayesianGenerator, __abstractmethods__=set())
+    def test_bad_mo_vocs(self):
+        vocs = VOCS(
+            **{
+                "variables": {"x1": [0, 1.0], "x2": [0, 10.0]},
+                "objectives": {"y1": "MINIMIZE"},
+                "constraints": {"c1": ["GREATER_THAN", 0.5]},
+                "constants": {"constant1": 1.0},
+            }
+        )
+        vocs2 = vocs.model_copy()
+        vocs2.objectives = {"y1": "MINIMIZE", "y2": "MAXIMIZE"}
+        with pytest.raises(ValueError):
+            gen = MultiObjectivePatchBayesianGenerator(
+                vocs=vocs, reference_point={"y1": 0.5, "y2": 0.5}
+            )
+
+        gen = MultiObjectivePatchBayesianGenerator(
+            vocs=vocs2, reference_point={"y1": 0.5, "y2": 0.5}
+        )
+        assert gen.supports_single_objective == False
+        with pytest.raises(ValueError):
+            gen.vocs = vocs

--- a/xopt/tests/generators/bayesian/test_mobo.py
+++ b/xopt/tests/generators/bayesian/test_mobo.py
@@ -9,7 +9,6 @@ import yaml
 from botorch.acquisition.multi_objective.logei import (
     qLogNoisyExpectedHypervolumeImprovement,
 )
-from pydantic import ValidationError
 
 from xopt.base import Xopt
 from xopt.errors import XoptError
@@ -31,7 +30,7 @@ class TestMOBOGenerator:
         vocs.objectives.update({"y2": "MINIMIZE"})
         reference_point = {"y1": 1.5, "y2": 1.5}
 
-        gen = MOBOGenerator(vocs=vocs, reference_point=reference_point)
+        MOBOGenerator(vocs=vocs, reference_point=reference_point)
 
         # test bad reference point
         with pytest.raises(XoptError):

--- a/xopt/tests/generators/bayesian/test_mobo.py
+++ b/xopt/tests/generators/bayesian/test_mobo.py
@@ -1,18 +1,22 @@
+import json
 from copy import deepcopy
 
 import numpy as np
 import pandas as pd
 import pytest
 import torch
+import yaml
 from botorch.acquisition.multi_objective.logei import (
     qLogNoisyExpectedHypervolumeImprovement,
 )
 from pydantic import ValidationError
 
 from xopt.base import Xopt
+from xopt.errors import XoptError
 from xopt.evaluator import Evaluator
 from xopt.generators.bayesian.mobo import MOBOGenerator
 from xopt.numerical_optimizer import GridOptimizer
+from xopt.pydantic import remove_none_values
 from xopt.resources.test_functions.tnk import (
     evaluate_TNK,
     tnk_reference_point,
@@ -27,11 +31,25 @@ class TestMOBOGenerator:
         vocs.objectives.update({"y2": "MINIMIZE"})
         reference_point = {"y1": 1.5, "y2": 1.5}
 
-        MOBOGenerator(vocs=vocs, reference_point=reference_point)
+        gen = MOBOGenerator(vocs=vocs, reference_point=reference_point)
 
         # test bad reference point
-        with pytest.raises(ValidationError):
+        with pytest.raises(XoptError):
             MOBOGenerator(vocs=vocs, reference_point={})
+
+    def test_round_trip(self):
+        vocs = deepcopy(TEST_VOCS_BASE)
+        vocs.objectives.update({"y2": "MINIMIZE"})
+        reference_point = {"y1": 1.5, "y2": 1.5}
+        gen = MOBOGenerator(vocs=vocs, reference_point=reference_point)
+        dump = gen.json()
+        gen2 = MOBOGenerator(vocs=vocs, **json.loads(dump))
+        assert dump == gen2.json()
+        assert gen.model_dump() == gen2.model_dump()
+        dump = gen.yaml()
+        gen2 = MOBOGenerator(vocs=vocs, **remove_none_values(yaml.safe_load(dump)))
+        assert dump == gen2.yaml()
+        assert gen.model_dump() == gen2.model_dump()
 
     def test_script(self):
         evaluator = Evaluator(function=evaluate_TNK)

--- a/xopt/tests/generators/bayesian/test_turbo.py
+++ b/xopt/tests/generators/bayesian/test_turbo.py
@@ -90,9 +90,12 @@ class TestTurbo(TestCase):
         gen.from_dict(gen_dict | {"vocs": test_vocs})
 
         # test invalid turbo controller type
+        test_vocs_2 = test_vocs.copy()
+        test_vocs_2.objectives = {}
+        test_vocs_2.observables = ["o1"]
         with pytest.raises(ValueError):
             gen = BayesianExplorationGenerator(
-                vocs=test_vocs, turbo_controller="OptimizeTurboController"
+                vocs=test_vocs_2, turbo_controller="OptimizeTurboController"
             )
 
     def test_get_trust_region(self):

--- a/xopt/tests/test_generator.py
+++ b/xopt/tests/test_generator.py
@@ -2,8 +2,8 @@ import json
 from copy import deepcopy
 
 import pytest
-from pydantic import ValidationError
 
+from xopt.errors import VOCSError
 from xopt.generator import Generator
 from xopt.generators import (
     get_generator,
@@ -34,7 +34,7 @@ class TestGenerator:
         test_vocs = deepcopy(TEST_VOCS_BASE)
         test_vocs.objectives.update({"y2": "MINIMIZE"})
 
-        with pytest.raises(ValidationError):
+        with pytest.raises(VOCSError):
             PatchGenerator(vocs=test_vocs)
 
     @pytest.mark.parametrize("name", list_available_generators())

